### ISSUE TITLE
Created Zeroed Pulse shapes

### ISSUE THREAD

**Issue by taalexander** [Qiskit/qiskit-terra#3742]
### What is the expected enhancement?
Currently, several pulses are implicitly zeroing the ends of the pulses to removed initial and final discontinuities for pulses that are settable with the flag `zero_ends` which is True. Naively reading the pulse name this would be unexpected behavior. This flag should be removed and the functionality should be split into two different pulses,eg.,  `ZeroedGaussian` and `Gaussian`. Similarily parametric pulse implementations should be made for these pulses.

- [ ] Split pulses with `zero_ends` attributes into two pulse classes
- [ ] Create parametric pulses for all split pulses

**Comment by taalexander** [comment]
> I think it's a good approach for generalization, especially for positive functions, but it doesn't solve the discontinuity issue when g(t)=0 which in this case is at t = width.

My apologies for not grasping this problem earlier, you are correct. I don't have much of an excuse other than not doing my due diligence 😦 . We should not touch the derivative component as the zeroing is a constant normalization term and taking the derivative of this term will result in no contribution to the derivative component. See equations (5) and (10) in the [original DRAG paper](https://arxiv.org/pdf/0901.0534.pdf).

**Comment by taalexander** [comment]
@eembees are you interested in this issue?

**Comment by eembees** [comment]
Yes, I'll start working over the weekend. 

> It may be possible to transform the derivative component to account for its finite duration and initial discontinuity but I believe this would be a non-trivial transformation.
The only way I see us being able to zero the edges of the derivative would be to multiply it by the `gaussian_square` pulse (also zeroed). I think that would be the best (if not only) way to keep it continuous without touching the inflection point in the center.

**Comment by taalexander** [comment]
My worry is that this transformation would ruin (or at least unpredictably modify) the assumptions that the DRAG pulse was derived under. I think to do it right we would have to undertake an analytical/numerical on the impacts.

**Comment by taalexander** [comment]
I should add that I am still uncertain about splitting the pulses into `Gaussian` and `ZeroedGaussian`. I worry that people will default to `Gaussian` or `DRAG` when `ZeroedDRAG` and `ZeroedGaussian` would almost always be better.

Any thoughts @lcapelluto, @eggerdj, @ajavadia, @dcmckayibm?

**Comment by ajavadia** [comment]
Can’t it be an option of the pulse builder to zero or not

**Comment by taalexander** [comment]
It currently is an option for the pulses. We do not want to support optional parameters in `ParametricPulse`.

**Comment by DanPuzzuoli** [comment]
I was just working on the docs for `discrete.py` and came across this issue (@lcapelluto pointed me to the _literal_ github issue). Let me know if this discussion has been taken further outside of these comments @eembees and @taalexander (and perhaps my comments are obsolete). 

I think that whatever the functionality is, it should be consistent. Currently, e.g.: 
- the `drag` function is inconsistent, in that if you call it with `zero_ends == True`, it returns samples from the function `g(x) + 1j * beta * h(x)`, where `g(x)` is the zeroed gaussian, but `h(x)` is the derivative of the _non-zeroed_ gaussian
-  So, the current `drag` function is not actually returning something satisfying the drag conditions. One response to this is that these things are equivalent up to a scaling which is absorbable into `beta`, but I think the current behaviour is "unpredictable" from a user perspective

So, I would suggest: 
- Wherever the discussion on having proper 'zeroing' for the derivatives goes, there should at least be a function that returns the derivative of the zeroed guassian; i.e. the "derivative of (zeroed gaussian)" as opposed to the "zeroed (derivative of gaussian)". 

To this end:
- The `_fix_gaussian_width` function maps each function output value `y` to:
`amp*(y - zero_offset)/(amp-zero_offset)`  (i.e. it is not just a constant offset)
- So, having a "derivative of (zeroed gaussian)" function requires scaling the derivative portion of the output of `gauss_deriv` by `amp/(amp-zero_offset)`

It may be a bit awkward but if "zeroing" functionality is to be provided, for full user control there needs to be multiple levels of "zeroing" options:
- drag with no zeroing
- drag with zeroing of the gaussian but not the derivative (but with consistent derivative as per the above suggestion), and
- drag with both the gaussian _and_ its derivative zeroed (consistently, so drag condition is preserved by whatever transformations are performed)

**Comment by taalexander** [comment]
@DanPuzzuoli 

> the drag function is inconsistent, in that if you call it with zero_ends == True, it returns samples from the function g(x) + 1j * beta * h(x), where g(x) is the zeroed gaussian, but h(x) is the derivative of the non-zeroed gaussia

The derivative of the zeroed gaussian is the same as the nonzeroed gaussian since it is a constant offset. The amplitude is also being rescaled for the gaussian derivative component properly see https://github.com/Qiskit/qiskit-terra/blob/0e2102057ad0d56eea50de132d69030c603df20a/qiskit/pulse/pulse_lib/continuous.py#L135 and https://github.com/Qiskit/qiskit-terra/blob/0e2102057ad0d56eea50de132d69030c603df20a/qiskit/pulse/pulse_lib/continuous.py#L343.

1. drag with no zeroing
2. drag with zeroing of the gaussian but not the derivative (but with consistent derivative as per the above suggestion),
3. drag with both the gaussian and its derivative zeroed (consistently, so drag condition is preserved by whatever transformations are performed, as already discussed above)


I understand why 1. and 3. are needed, but I don't understand the use case of 2.

**Comment by DanPuzzuoli** [comment]
@taalexander 

Yes sorry my bad, I thought I had been very careful before posting the comment, but you are totally correct. 

What I meant by the 1, 2, 3 cases has 2 as what is currently being done when `zero_ends == True`: it is enforced that the output gaussian starts and ends at 0, but the derivative can have non-zero starting/ending points.

Case 3 corresponds to what we just discussed in person; it would correspond to a pulse shape that starts and ends at 0 that also has derivative starting/ending at 0. (To recap/record our discussion: it isn't possible to do this using the set of functions parameterized in Equation 5 of the drag paper. So, if one wants to enforce that the function _and_ its first derivative start/end at 0, it is necessary to consider a broader class of functions [e.g. a linear combo of gaussians with different widths].)

**Comment by taalexander** [comment]
Another option would be to add the parameter `B` which would be the offset of a gaussian, this could be 0 by default and then set by the `drag` pulse function to appropriately zero the pulse if desired.

**Comment by eembees** [comment]
> Another option would be to add the parameter `B` which would be the offset of a gaussian, this could be 0 by default and then set by the `drag` pulse function to appropriately zero the pulse if desired.

And then add a flag on whether that should propagate onto the derivative?  Or do you still think we should add a separate set of classes for DRAG pulses? I'm a bit confused on exactly what direction this should go ATM.

**Comment by taalexander** [comment]
It might look something like
```
Gaussian(duration, amp, b, sigma)
```
```
gaussian(duration, amp, b, sigma, zero_ends=True) -> Gaussian(duration, adjust(amp), adjust(b), sigma)
gaussian(duration, amp, b, sigma, zero_ends=False) -> Gaussian(duration, amp, 0, sigma)
```
I'm not really certain currently. Perhaps we should close this and reopen at a later date when it is no longer a hypothetical problem?

**Comment by HuangJunye** [comment]
I was looking through "good first issues" to work on and found this one. But it seems this new feature request is already addressed by #4640 as mentioned in #4317

**Comment by taalexander** [comment]
Yes, I believe so @HuangJunye. Thank you for catching.
